[ONNX] Support optional type (#68793)

BowenBao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73284

Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguel@microsoft.com>
Co-authored-by: neginraoof <neginmr@utexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

BowenBao committed:
[ONNX] Support optional type (#68793)

Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: BowenBao <bowbao@microsoft.com>
Co-authored-by: neginraoof <neginmr@utexas.edu>

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: BowenBao <bowbaomicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: BowenBao <bowbaomicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: BowenBao <bowbaomicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

[ghstack-poisoned]

malfet commented on torch/csrc/jit/python/python_arg_flatten.cpp:
```diff
@@ -68,7 +72,7 @@ void flatten_rec(PyObject* obj, ParsedArgs& args) {
     args.desc.metadata.emplace_back(var);
     args.desc.structure.push_back(D::Variable);
-  } else if (strcmp(THPUtils_typename(obj), "NoneType") == 0) {
+  } else if (PyNone_Check(obj)) {
```
This change does not seem to be specific to a PR, is it? If so, can it be moved to a separate one?

BowenBao replied:
cc original author @garymm

garymm replied:
@BowenBao do you want me to open another PR in `onnx_ms_1` that reverts the changes in this file? Or how shall I proceed?

BowenBao replied:
@garymm is it indeed not specific to this PR? Personally I do not prefer separation, since this PR already touches a lot of other files outside of ONNX as well. 

If there is a need to separate, I think the process could be you directly work on the branch of this PR, and push new commits to this branch. The separated change will be another PR submitted against master.

garymm replied:
I think it's an unrelated clean-up, but since we're getting a TorchScript code owner to review the PR anyways, I don't see any value in removing this from the PR.

Marking this thread resolved. If the TorchScript owner asks for this to be separated I can do so.

suo commented:
sorry, missed the ping. @eellison do you mind looking over the JIT changes

eellison commented on aten/src/ATen/core/interned_strings.h:
```diff
@@ -1,9 +1,9 @@
 #pragma once
-#include <vector>
+#include <algorithm>
```
Can you revert the formatting changes here in this PR or stack them ?

BowenBao replied:
sure, reverted the formatting part.

eellison commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -445,34 +445,29 @@ static TypePtr inferShapeAndTypeForInput(
     auto type = getTensorType(s_iter->toTensor(), complete);
     s_iter++;
     return type;
+  } else if (auto optional_type = input_type->cast<OptionalType>()) {
```
Why do we need the changes here ?

BowenBao replied:
To include input with OptionalType. I believe previously they are ignored.

eellison commented on torch/csrc/jit/python/script_init.cpp:
```diff
@@ -492,16 +487,19 @@ static void setInputTensorTypes(
     TORCH_INTERNAL_ASSERT(input_values.size() == param_count_list.size());
   }
   for (auto v : input_values) {
-    AT_ASSERT(s_iter != stack.end());
```
Why do we need changes here ?

eellison reviewed (commented):
Couple questions

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

BowenBao commented:
Thanks @eellison for providing comments.

@garymm I updated the part related to reverting formatting change. I tried to answer the comments to the best of my knowledge. Please feel free to add more comments.

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34625646).

naor2013 commented:
@BowenBao I have a question regarding this pull request.
If I need the __is__ operator you added for "is None" use cases, can I just use the opset you added as a custom opset in PyTorch 1.11 or will it not work without the other changes you added to PyTorch?

garymm commented:
@naor2013 nothing in this PR will be available in PyTorch 1.11.
You can modify an ONNX model to make use of optional types using the onnx Python API or any protocol buffer API.

naor2013 commented:
@garymm my question is basically, can I take the __is_ opset added in symbolic_opset15.py in this commit and use it as a custom opset or will it not work without all the other changes in this PR outside of that file?

thiagocrepaldi commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,100 @@
+# Owner(s): ["module: onnx"]
```
did you check whether the pipeline is calling this file?

garymm replied:
I think it is not, thanks for pointing out.  I don't want to touch this PR while it's being imported by Meta. I ran this test locally and confirms it passes, and after this is merged we can open another PR to run this automatically.

BowenBao replied:
Tests under `test/onnx` by default runs in shard 1.

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

BowenBao commented:
@malfet rebased again to resolve conflicts. Please import

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

justinchuby commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,100 @@
+            dynamic_axes={"y": {0: "y0", 1: "y1"}}, input_names=["y"])
+
+
+instantiate_parametrized_tests(TestOptionalOutput)
```
Is this performed by https://github.com/pytorch/pytorch/blob/a997046017aa9f97213f6bd50bc152a6445e8086/torch/testing/_internal/common_utils.py#L240 already? (Seen in instantiate_parametrized_tests's doc)

BowenBao replied:
We'll discuss these offline, more likely update it in a follow up PR.

justinchuby commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,100 @@
+                    if attr.name in ("then_branch", "else_branch"):
+                        self.assertEqual(expected_output_type, attr.g.output[0].type)
+
+    def test_uninitialized_optional(self):
```
I would include the expected behavior as part of the test method name.

justinchuby commented on test/onnx/test_pytorch_onnx_no_runtime.py:
```diff
@@ -0,0 +1,100 @@
+
+
+class TestOptionalOutput(unittest.TestCase):
+    # TODO: Move these tests to test_pytorch_onnx_onnxruntime once
```
I would add your username after the todos so that readers know the right contact for this todo. (same for other todo's)

```suggestion
    # TODO(BowenBao): Move these tests to test_pytorch_onnx_onnxruntime once
```

garymm replied:
blame can be used if needed. But really the comments should either be self-explanatory or contain a link that has more details. In neither case should we rely on the memory of a human.

BowenBao committed:
Update on "[ONNX] Support optional type (#68793)"


Some important ops won't support optional type until opset 16,
so we can't fully test things end-to-end, but I believe this should
be all that's needed. Once ONNX Runtime supports opset 16,
we can do more testing and fix any remaining bugs.

Co-authored-by: garymm <garymiguelmicrosoft.com>
Co-authored-by: neginraoof <neginmrutexas.edu>

Differential Revision: [D34625646](https://our.internmc.facebook.com/intern/diff/D34625646)

[ghstack-poisoned]

malfet commented:
@malfet has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34625646).

malfet commented:
This change once again regresses some of the internal model conversion tests, even though they are pinned to OpSet9 and as such should not be affected by the change, are they?

garymm commented on torch/onnx/utils.py:
```diff
@@ -440,37 +439,51 @@ def _decide_constant_folding(do_constant_folding, operator_export_type, training
+                param = sig.parameters[optional_arg]
+                if param.default != param.empty:
+                    args_list.append(param.default)
+        args = tuple(args_list)
```
How's about:
`args = args_list if isinstance(args, list) else tuple(args_list)`
?

Thanks!

malfet committed:
Update torch/onnx/utils.py

datumbox commented:
@BowenBao @malfet @eellison This PR is likely to have broken TorchVision. Please see https://github.com/pytorch/vision/issues/5971 for details. Can you please revert?

garymm commented:
I'll look at this today. It took us 3 months to get this PR merged so even though it's generally bad practice, I'd really like to fix forward if possible.

datumbox commented:
@garymm Thanks for coming back to me. I understand it, if you want to explore fixing the issue rather than reverting the whole PR. Just please prioritise this because TorchVision's ONNX export is broken for the detection models and we are dangerous close to the release. I'm also concerned that the longer we leave this in, the harder it will be to revert it if a fix can't be completed soon.

@malfet @eellison Any thoughts/concerns on your side?
